File: src/main/java/tmi/recipe/parser/ReconstructorParser.java
package tmi.recipe.parser;

import arc.func.Boolf;
import arc.func.Cons2;
import arc.struct.ObjectMap;
import arc.struct.Seq;
import mindustry.ctype.UnlockableContent;
import mindustry.type.UnitType;
import mindustry.world.Block;
import mindustry.world.blocks.units.Reconstructor;
import mindustry.world.consumers.Consume;
import tmi.recipe.Recipe;
import tmi.recipe.RecipeType;

public class ReconstructorParser extends ConsumerParser<Reconstructor>{
  @Override
  public boolean isTarget(Block content) {
    return content instanceof Reconstructor;
  }

  @Override
  public Seq<Recipe> parse(Reconstructor reconstructor) {
    Seq<Recipe> res = new Seq<>();
    for (UnitType[] upgrade : reconstructor.upgrades) {
      Recipe recipe = new Recipe(RecipeType.factory);
      recipe.setBlock(getWrap(reconstructor));
      recipe.setTime(reconstructor.constructTime);
      recipe.addMaterial(getWrap(upgrade[0]), 1);
      recipe.addProduction(getWrap(upgrade[1]), 1);

      registerCons(recipe, reconstructor.consumers);

      res.add(recipe);
    }

    return res;
  }
}

File: src/main/kotlin/tmi/recipe/parser/ReconstructorParser.kt
package tmi.recipe.parser

import arc.struct.Seq
import mindustry.world.Block
import mindustry.world.blocks.units.Reconstructor
import tmi.recipe.Recipe
import tmi.recipe.RecipeType

open class ReconstructorParser : ConsumerParser<Reconstructor>() {
  override fun isTarget(content: Block): Boolean {
    return content is Reconstructor
  }

  override fun parse(content: Reconstructor): Seq<Recipe> {
    val res = Seq<Recipe>()
    for (upgrade in content.upgrades) {
      val recipe = Recipe(
        recipeType = RecipeType.factory,
        ownerBlock = +content,
        craftTime = content.constructTime,
      )

      recipe.addMaterialInteger(+upgrade[0], 1)
      recipe.addProductionInteger(+upgrade[1], 1)

      registerCons(recipe, *content.consumers)

      res.add(recipe)
    }

    return res
  }
}
